kore/src/Kore/Equation/Application.hs

```diff
@@ -192,10 +194,13 @@ attemptEquation sideCondition termLike equation =
               & whileMatch
           (equation', predicate) <-
               applyAndSelectMatchResult matchResults
-          return
-              ( equation'
-              , makeAndPredicate predicate matchPredicate
-              )
+
+          assert (not . hasFreeEquationVariableName $ equation')
+              $ assert (not . hasFreeEquationVariableName $ predicate)
```

Comment:
Could this be factored out? I think it's a little too verbose and it takes longer to read than necessary.